include/targets/cuda/reduction_kernel.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <target_device.h>
+#include <reduce_helper.h>
+
+namespace quda {
+
+  /**
+     @brief This class is derived from the arg class that the functor
+     creates and curries in the block size.  This allows the block
+     size to be set statically at launch time in the actual argument
+     class that is passed to the kernel.
+  */
+  template <int block_size_x_, int block_size_y_, typename Arg_> struct ReduceKernelArg : Arg_ {
+    using Arg = Arg_;
+    static constexpr int block_size_x = block_size_x_;
+    static constexpr int block_size_y = block_size_y_;
+    ReduceKernelArg(const Arg &arg) : Arg(arg) { }
+  };
+
+  template <template <typename> class Transformer, typename Arg, bool grid_stride = true>
+  __forceinline__ __device__ void Reduction2D_impl(const Arg &arg)
+  {
+    using reduce_t = typename Transformer<Arg>::reduce_t;
+    Transformer<Arg> t(arg);
+
+    auto idx = threadIdx.x + blockIdx.x * blockDim.x;
+    auto j = threadIdx.y;
```

Comment:
Similar to the comment below, I'm guessing there's a reason why this isn't `+ blockIdx.y * blockDim.y`, please add a comment on it since on first pass it looks like a typo.